vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/AbstractColumn.java

```diff
@@ -311,23 +325,20 @@ protected boolean hasSortingIndicators() {
         return sortingIndicators;
     }
 
-    /*
-     * The original header template is needed for when sorting is enabled or
-     * disabled in a column.
-     */
-    protected void setBaseHeaderTemplate(String headerTemplate) {
-        rawHeaderTemplate = headerTemplate;
-    }
-
     /*
      * Adds the sorting webcomponent markup to an existing template.
      */
     protected String addGridSorter(String templateInnerHtml) {
         String escapedColumnId = HtmlUtils
                 .escape(getBottomLevelColumn().getInternalId());
 
-        String textContent = org.jsoup.Jsoup.parse(templateInnerHtml).text();
-        String sortBy = textContent.isBlank() ? ""
+        String textContent = headerComponent != null
+                ? headerComponent.getElement().getTextRecursively()
+                : headerText;
+        if (textContent != null) {
+            textContent = HtmlUtils.escape(textContent);
+        }
```

Comment:
For header components use `getTextRecursively` to get a text for the aria-label.

Also escape the text, which might be something that was done before by parsing and extracting the text through Jsoup.